plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -2909,147 +2863,143 @@ void producer_plugin_impl::start_write_window() {
             self->switch_to_write_window();
           } else if ( self ) {
-             self->_ro_trx_exec_tasks_fut.clear();
+             self->_ro_exec_tasks_fut.clear();
           }
        }));
 }
 
-// Called from a read only trx thread. Run in parallel with app and other read only trx threads
-bool producer_plugin_impl::read_only_trx_execution_task(fc::time_point start) {
-   // We have 4 ways to break out the while loop:
+// Called from a read only thread. Run in parallel with app and other read only threads
+bool producer_plugin_impl::read_only_execution_task(uint32_t pending_block_num) {
+   // We have 3 ways to break out the while loop:
    // 1. pass read window deadline
-   // 2. Net_plugin receives a block
-   // 3. No more transactions in the read-only trx queue
-   // 4. A transaction execution is exhaused
-   ro_trx_t trx;
-   while ( _ro_trx_queue.pop_front(trx) ) {
-      // If the queue is empty, pop_front() waits on condition variable, and returns false
-      // when and only when all tasks must exit (i.e queue is empty and all tasks are idle, or
-      // we have reached the end of the read window, or net plugin received a block)
-
-      auto retry = process_read_only_transaction( trx.trx, trx.next, start );
-      if ( retry ) {
-         _ro_trx_queue.push_front(std::move(trx));
-         // Do not schedule new execution
+   // 2. net_plugin receives a block
+   // 3. no read-only tasks to execute
+   while ( fc::time_point::now() < _ro_window_deadline && _received_block < pending_block_num ) {
+      bool more = app().executor().execute_highest_read_only(); // blocks until all read only threads are idle
+      if ( !more ) {
          break;
       }
    }
 
    // If all tasks are finished, do not wait until end of read window; switch to write window now.
-   if ( --_ro_num_active_trx_exec_tasks == 0 ) {
-      // Do switching on app thread to serialize
-      app().executor().post( priority::high, exec_queue::read_only_trx_safe, [self=this]() {
-         self->_ro_trx_exec_tasks_fut.clear();
+   if ( --_ro_num_active_exec_tasks == 0 ) {
+      // Needs to be on read_only because that is what is being processed until switch_to_write_window().
+      app().executor().post( priority::high, exec_queue::read_only, [self=this]() {
+         self->_ro_exec_tasks_fut.clear();
+         // will be executed from the main app thread because all read-only threads are idle now
          self->switch_to_write_window();
       } );
+      // last thread post any exhausted back into read_only queue with slightly higher priority (low+1) so they are executed first
+      ro_trx_t t;
+      while( _ro_exhausted_trx_queue.pop_front(t) ) {
+         app().executor().post(priority::low+1, exec_queue::read_only, [this, trx{std::move(t.trx)}, next{std::move(t.next)}]() mutable {
+            push_read_only_transaction( std::move(trx), std::move(next) );
+         } );
+      }
```

Comment:
```suggestion
      repost_exhausted_transactions();
```

note: remove unused `deadline` parameter in `repost_exhausted_transactions`

Reply:
It should be used and it is now.

Reply:
Why are we checking the deadline in `repost_exhausted_transactions`? This moving from a queue to another one is pretty much instant if I understand correctly. It is not like we are processing anything.

Reply:
It is just a safe guard in case somehow the queue gets extra large.  All `while` loops in start block call `should_interrupt_start_block()`. I've updated this to do the same.

Reply:
`remove_expired_trxs` uses `pending_block_time` not `preprocess_deadline` for determination of which trxs to expire.
The `preprocess_deadline` is exclusively used for determining if too much time has passed in `start_block`.

Reply:
Oh, yes, got it, my bad, thanks!